proposals/p2138.md

```diff
@@ -0,0 +1,51 @@
+# Checked and template generic terminology
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2138)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Proposal](#proposal)
+-   [Rationale](#rationale)
+
+<!-- tocstop -->
+
+## Abstract
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
+
+## Problem
+
+The C++ community does "generic programming" using templates in C++, and so
+thinks of templates as a mechanism for implementing generics. In Carbon,
+templates and other generic features are more similar than they are different,
+so it is worthwhile to have an easy way to talk about things that apply to both.
+
+## Proposal
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
+
+The [generics overview](/docs/design/generics/overview.md), for example,
+currently demonstrates the current approach of using "generics" and "templates".
+The [generics section of the design overview](/docs/design/README.md#generics)
+demonstrates the use of the proposed new terminology.
```

Comment:
This sounds like an alternative considered: what were the advantages and disadvantages of the other approach?

Reply:
I think these documents are examples of the new proposed terminology.

We could add an alternatives considered section that talks about the status quo if useful -- I think that's the only alternative really considered here. But I worry it would be a bit redundant as the problem statement already somewhat talks about that.

Reply:
I've written a bit of framing to clarify that these are just intended to show examples of before and after.

Reply:
FWIW, I think I'm coming from a different perspective than both of you. I think there was actually an advantage to just saying "templates" instead of "template generics" for C++ familiarity (e.g., https://en.cppreference.com/w/cpp/language/templates exclusively says "template" and doesn't mention "generic", and I actually think "template generics" could get confused with C++ constraints/concepts due to how they mimic generics in other languages). Maybe "templates" and "checked generics" would be better for C++ developer familiarity?

So to note a few points:

- "Problems" describes the downsides of the current approach. Alternatives should also consider advantages.
- Rationale notes that the new naming improves familiarity for C++ developers. But how familiar are C++ developers with "template generics" and "typed generics" as those specific terms?
- In a design, you would link to "Alternatives considered". The lack of "Alternatives considered" means it would only be linked as a reference. (I thought we had a proposal describing link structuring, but I'm not finding it right now)

But again, maybe leads are fine with all the above.